src/graph.py

```diff
@@ -0,0 +1,286 @@
+import collections
+import logging
+import time
+from typing import Optional
+
+import ccxt
+
+log = logging.getLogger(__name__)
+
+
+class PricePath:
+    def __init__(
+        self,
+        exchanges: Optional[list[str]] = None,
+        gdict: Optional[dict] = None,
+        cache: Optional[dict] = None,
+    ):
+        if exchanges is None:
+            exchanges = []
+        if gdict is None:
+            gdict = {}
+        if cache is None:
+            cache = {}
+
+        self.gdict = gdict
+        self.cache = cache
+
+        # Saves the priority for a certain path so that bad paths can be skipped.
+        self.priority: collections.defaultdict[str, int] = collections.defaultdict(int)
+        allpairs: list[tuple[str, str, str, str]] = []
+
+        for exchange_id in exchanges:
+            exchange_class = getattr(ccxt, exchange_id)
+            exchange = exchange_class()
+            markets = exchange.fetch_markets()
+            assert isinstance(markets, list)
+
+            if exchange.has["fetchOHLCV"]:
+                allpairs.extend(
+                    [(i["base"], i["quote"], exchange_id, i["symbol"]) for i in markets]
+                )
+            else:
+                logging.warning(
+                    f"{exchange.name} does not support fetch ohlcv. "
+                    f"Ignoring exchange and {len(markets)} pairs."
+                )
+
+        # Remove duplicate pairs.
+        # TODO It might be faster to create it directly as set.
+        #      Is it even necessary to convert it to a list?
+        allpairs = list(set(allpairs))
+        # print("Total Pairs to check:", len(allpairs))
+
+        # Sorting by `symbol` to have the same result on every run due to the set.
+        allpairs.sort(key=lambda x: x[3])
+
+        for base, quote, exchange, symbol in allpairs:
+            self.addVertex(base)
+            self.addVertex(quote)
+            self.addEdge(
+                base, quote, {"exchange": exchange, "symbol": symbol, "inverted": False}
+            )
+            self.addEdge(
+                quote, base, {"exchange": exchange, "symbol": symbol, "inverted": True}
+            )
+
+    def edges(self):
+        return self.findedges()
+
+    # Find the distinct list of edges
+
+    def findedges(self):
+        edgename = []
+        for vrtx in self.gdict:
+            for nxtvrtx in self.gdict[vrtx]:
+                if {nxtvrtx, vrtx} not in edgename:
+                    edgename.append({vrtx, nxtvrtx})
+        return edgename
+
+    def getVertices(self):
+        return list(self.gdict.keys())
+
+    # Add the vertex as a key
+    def addVertex(self, vrtx):
+        if vrtx not in self.gdict:
+            self.gdict[vrtx] = []
+
+    def addEdge(self, vrtx1, vrtx2, data):
+        if vrtx1 in self.gdict:
+            self.gdict[vrtx1].append((vrtx2, data))
+        else:
+            self.gdict[vrtx1] = [vrtx2]
```

Comment:
Thats actually a bug, that never executes, because the verticies are inserted before the edges thus creating the entiry needed in the dict.
fix would be:
```
        if vrtx1 in self.gdict:
            self.gdict[vrtx1].append((vrtx2, data))
        else:
            self.gdict[vrtx1] = [(vrtx2, data)]
```